File: problems/0679_24_game/0679_24_game.java
class Solution {
    private static final double EPSILON = 1e-6;
    private static final double TARGET = 24.0;

    public boolean judgePoint24(int[] nums) {
        List<Double> list = new ArrayList<>();
        for (int i : nums) {
            list.add((double) i);
        }
        return dfs(list);
    }

    private boolean dfs(List<Double> list) {
        if (list.size() == 1) {
            if (Math.abs(list.get(0) - TARGET) < EPSILON) {
                return true;
            }
            return false;
        }

        for (int i = 0; i < list.size(); i++) {
            for (int j = i + 1; j < list.size(); j++) {
                for (double c : compute(list.get(i), list.get(j))) {
                    List<Double> next = new ArrayList<>();
                    next.add(c);
                    for (int k = 0; k < list.size(); k++) {
                        if (k == j || k == i)
                            continue;
                        next.add(list.get(k));
                    }
                    if (dfs(next))
                        return true;
                }
            }
        }
        return false;

    }

    private List<Double> compute(double a, double b) {
        List<Double> res = new ArrayList<>();
        res.add(a + b);
        res.add(a - b);
        res.add(b - a);
        res.add(a * b);
        res.add(a / b);
        res.add(b / a);
        return res;
    }
}

File: problems/0679_24_game/0679_24_game.py
class Solution:
    EPSILON = 1e-6
    TARGET = 24.0
    def judgePoint24(self, cards: List[int]) -> bool:
        nums = [float(n) for n in cards]
        return self.dfs(nums)
    def dfs(self, nums: list[float]) -> bool:
        if len(nums) == 1:
            return abs(nums[0] - self.TARGET) < self.EPSILON
        n = len(nums)
        for i in range(n):
            for j in range(i + 1, n):
                a, b = nums[i], nums[j]
                for c in self.compute(a, b):
                    next_nums = [c] + [nums[k] for k in range(n) if k != i and k != j]
                    if self.dfs(next_nums):
                        return True
        return False
    
    def compute(self, a: float, b: float):
        res = [a + b, a - b, b - a, a * b]
        if abs(b) > self.EPSILON:
            res.append(a / b)
        if abs(a) > self.EPSILON:
            res.append(b / a)
        return res